openedx_learning/lib/fields.py

```diff
@@ -117,3 +161,123 @@ def manual_date_time_field():
+                + "Please use a db_collations dict instead."
+            )
+
+        super().__init__(*args, **kwargs)
+        self.db_collations = db_collations or {}
+
+        # This is part of a hack to get this to work for Django < 4.1. Please
+        # see comments in the db_collation method for details.
+        self._vendor = None
+
+    @property
+    def db_collation(self):
+        """
+        Return the db_collation, understanding that it varies by vendor.
+
+        This method is a hack for Django 3.2 compatibility and should be removed
+        after we move to 4.2.
+
+        Description of why this is hacky:
+
+        In Django 4.2, the schema builder pulls the collation settings from the
+        field using the value returned from the ``db_parameters`` method, and
+        this does what we want it to do. In Django 3.2, field.db_parameters is
+        called, but any collation value sent back is ignored and the code grabs
+        the value of db_collation directly from the field:
+
+        https://github.com/django/django/blob/stable/3.2.x/django/db/backends/base/schema.py#L214-L224
+
+        But this call to get the ``field.db_collation`` attribute happens almost
+        immediately after the ``field.db_parameters`` method call. So our
+        fragile hack is to set ``self._vendor`` in the ``db_parameters`` method,
+        using the value we get from the connection that is passed in there.
+
+        This method, the corresponding setter, and all references to
+        ``self._vendor`` should be removed after we've cut over to Django 4.2.
+        """
+        return self.db_collations.get(self._vendor)
+
+    @db_collation.setter
+    def db_collation(self, value):
+        """
+        Don't allow db_collation to be set manually (just ignore).
+
+        This can be removed when we move to Django 4.2.
+        """
+        pass
+
+    def db_parameters(self, connection):
+        """
+        Return database parameters for this field. This adds collation info.
+
+        We examine this field's ``db_collations`` attribute and return the
+        collation that maps to ``connection.vendor``. This will typically be
+        'mysql' or 'sqlite'.
+        """
+        db_params = models.Field.db_parameters(self, connection)
+
+        # Remove once we no longer need to support Django < 4.1
+        self._vendor = connection.vendor
+
+        # Now determine collation based on DB vendor (e.g. 'sqlite', 'mysql')
+        if connection.vendor in self.db_collations:
+            db_params["collation"] = self.db_collations[connection.vendor]
+
+        return db_params
+
+    def deconstruct(self):
+        name, path, args, kwargs = super().deconstruct()
+        if self.db_collations:
+            kwargs["db_collations"] = self.db_collations
+        return name, path, args, kwargs
+
+ 
+class MultiCollationCharField(MultiCollationMixin, models.CharField):
```

Comment:
@bradenmacdonald, @kdmccormick, @feanil: May I get a sanity check from one of you folks on whether this class (and the mixin above it that does all the work) makes sense?

Reply:
Makes sense to me!

Though btw, is there a compelling reason to run tests in sqlite? I generally feel like it's better to test as your run in prod, which would mean using MySQL for tests, which would avoid most of the concerns here.

Reply:
Maybe this is silly on my part, but I'd like to hold the door open for sqlite on the low end. I have this vague idea of a [light configuration that uses sqlite for everything](https://openedx.atlassian.net/wiki/spaces/~6193b016d5986c006a6460ab/pages/3750953071/Lightest+Plausible+LMS+Studio+Configuration).

Though I do agree that testing in MySQL would be helpful.